File: SemiSupervisedPosTagger/SemiSupervisedPosTagger/file_manager.cpp
//
// Created by Mohammad Sadegh Rasooli on 1/7/15.
// Copyright (c) 2015 Mohammad Sadegh Rasooli. All rights reserved.
//

#include "file_manager.h"
#include "iostream"
#include "set"
#include "unicode_manager.h"

using namespace std;
index_maps file_manager::create_indexMaps(string file_path,const string delim) {
	cout << "creating index maps..."<<flush;
	set<string> words;
	set<string> tags;

	// 0 for <start>
	// 1 for <end>
	int index = 2;
	int tag_size;
	unordered_map<string, int> string_dic;
	vector<string> reverse_dict;

	// first indexing the tags
	ifstream reader(file_path);
	if(!reader.good()){
		cout << "file "<< file_path<< " does not exit"<<endl;
		throw exception();
	}


	while (!reader.eof()) {
		string line;
		getline(reader, line);
		if (line.length() > 0) {
			istringstream iss(line);
			do {
				string sub;
				iss >> sub;
				if (sub.length() > 0) {
					string tag = sub.substr(sub.rfind(delim) + 1);
					tags.insert(tag);
				}
			} while (iss);
		}
	}
	reader.close();

	set<string>::iterator it;
	for (it = tags.begin(); it != tags.end(); ++it) {
		string tag = *it;
		if (string_dic.count(tag) <= 0) {
			string_dic[tag] = index++;
		}
	}
	tag_size=index;

	// then indexing other strings
	reader.open(file_path);
	while (!reader.eof()) {
		string line;
		getline(reader, line);
		if (line.length() > 0) {
			istringstream iss(line);
			do {
				string sub;
				iss >> sub;
				if (sub.length() > 0) {
					string word = sub.substr(0, sub.rfind(delim));
					words.insert(word);

					vector<string> prefixes = unicode_manager::prefixes(word, 4);
					for(int i=0;i<prefixes.size();i++) {
						words.insert(prefixes.at(i));
					}

					vector<string> suffixes = unicode_manager::suffixes(word, 4);
					for(int i=0;i<suffixes.size();i++) {
						words.insert(suffixes.at(i));
					}
				}
			} while (iss);
		}
	}


	for (it = words.begin(); it != words.end(); ++it) {
		string word = *it;
		if (string_dic.count(word) <= 0) {
			string_dic[word] = index++;
		}
	}

	vector<string> rv_str;
	vector<int> rv_int;
	for(auto kv:string_dic){
		rv_str.push_back(kv.first);
		rv_int.push_back(kv.second-2);
	}

	reverse_dict.push_back("<START>");
	reverse_dict.push_back("<STOP>");
	for(int i=2;i<string_dic.size();i++) {
			reverse_dict.push_back(rv_str.at(rv_int.at(i)));
	}

	cout << "done!\n"<<flush;

	return index_maps(string_dic, reverse_dict,tag_size);
}

vector<sentence> file_manager::read_sentences(string file_path, unordered_map<string, int> string_dict,const string delim) {
	vector<sentence> sentences;
	ifstream reader(file_path);

	if(!reader.good()){
		cout << "file "<< file_path<< " does not exist"<<endl;
		throw exception();
	}

	cout << "reading the file "<<file_path<<"..."<<flush;
	int count=0;
	while (!reader.eof()) {
		string line;
		getline(reader, line);
		if (line.length() > 0) {
			count++;
			if(count%100==0)
				cout <<count<<"..."<<flush;
			sentence sen=get_sentence_from_line(line, string_dict,delim);
			if(sen.length>0)
			sentences.push_back(sen);
		}
	}
	cout <<"done!"<<endl;

	return sentences;
}

sentence file_manager::get_sentence_from_line(const string line, unordered_map<string, int> string_dict,const string delim) {
	istringstream iss(line);
	vector<string> words;
	vector<string> tags;
	int len = 0;
	do {
		string sub;
		iss >> sub;
		if (sub.length() > 0) {
			int ind=sub.rfind(delim);
			if(ind>0) {
				string word = sub.substr(0, ind);
				string tag = sub.substr(ind + 1, sub.length());
				words.push_back(word);
				tags.push_back(tag);
				len++;
			}
		}
	} while (iss);
	return sentence(words, tags, len, string_dict);
}

File: src/SemiSupervisedPOSTagger/IO/FileManager.java
package SemiSupervisedPOSTagger.IO;

import SemiSupervisedPOSTagger.Structures.IndexMaps;
import SemiSupervisedPOSTagger.Structures.Sentence;

import java.io.BufferedReader;
import java.io.FileReader;
import java.nio.Buffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;

/**
 * Created by Mohammad Sadegh Rasooli.
 * ML-NLP Lab, Department of Computer Science, Columbia University
 * Date Created: 1/8/15
 * Time: 6:50 PM
 * To report any bugs or problems contact rasooli@cs.columbia.edu
 */

public class FileManager {
    public static ArrayList<Sentence> readSentences(String filePath, IndexMaps maps,String delim) throws Exception{
        System.out.print("reading sentences...");
        BufferedReader reader=new BufferedReader(new FileReader(filePath));
        String line;
        ArrayList<Sentence> sentences=new ArrayList<Sentence>();
        while((line=reader.readLine())!=null){
            if(line.trim().length()>0)
                sentences.add(new Sentence(line,maps,delim));
        }
        System.out.print("done!\n");
        return sentences;
    }

    public static IndexMaps createIndexMaps(String filePath, String delim,String clusterFile,String tagDictionaryPath, int brownSize) throws Exception{
        System.out.print("creating index maps...");
        HashMap<String, Integer> stringMap=new HashMap<String, Integer>();
        HashMap<String,Integer> clusterMap=new HashMap<String, Integer>();
        HashMap<Integer,Integer>[] clusterNMap=new HashMap[brownSize];
        for(int i=0;i<brownSize;i++)
            clusterNMap[i]=new HashMap<Integer, Integer>();

        BufferedReader reader=new BufferedReader(new FileReader(filePath));

        HashSet<String> tags = new HashSet<String>();
        HashSet<String> words=new HashSet<String>();

        String line;
        while((line=reader.readLine())!=null){
            String[] split=line.trim().split(" ");
            for(int i=0;i<split.length;i++){
                if(split[i].contains(delim)){
                    int delimIndex=split[i].lastIndexOf(delim);
                    String word=split[i].substring(0,delimIndex);

                    for(int p=0;p<Math.min(4,word.length());p++){
                        String prefix=word.substring(0,p+1);
                        String suffix=word.substring(word.length()-p-1);
                        words.add(prefix);
                        words.add(suffix);
                    }

                    String tag=split[i].substring(delimIndex+1);
                    tags.add(tag);
                    words.add(word);
                }
            }
        }

        // 0 and 1 are reserved for stop and start
        int index=2;
        stringMap.put("<<START>>",0);
        stringMap.put("<<STOP>>", 1);

        for(String t:tags){
            stringMap.put(t,index++);
        }

        if(clusterFile.length()>0){
            reader = new BufferedReader(new FileReader(clusterFile));
            while ((line = reader.readLine()) != null) {
                String[] spl = line.trim().split("\t");
                if (spl.length > 2) {
                    String cluster = spl[0];
                    String word=spl[1];
                    int clusterNum=index;

                    if (!stringMap.containsKey(cluster)) {
                        clusterMap.put(word,index);
                        stringMap.put(cluster, index++);
                    }else{
                        clusterNum= stringMap.get(cluster);
                        clusterMap.put(word,clusterNum);
                    }

                    for(int i=0;i<brownSize;i++) {
                        int prefId = index;
                        String prefix=cluster.substring(0,Math.min(i+1,cluster.length()));
                        if (!stringMap.containsKey(prefix)) {
                            stringMap.put(prefix, index++);
                        } else {
                            prefId = stringMap.get(prefix);
                        }
                        clusterNMap[i].put(clusterNum,prefId);
                    }

                }
            }
        }
        
        for(String w:words){
            if(!stringMap.containsKey(w))
                stringMap.put(w,index++);
            if(!stringMap.containsKey(w.toLowerCase()))
               stringMap.put(w.toLowerCase(),index++);
        }

        System.out.println(stringMap.size());
        String[] reversedMap=new String[stringMap.size()];
        for(String k:stringMap.keySet()){
            reversedMap[stringMap.get(k)]=k;
        }

        int tagSize=tags.size()+2;
        System.out.print("done!\n");

        HashMap<Integer,HashSet<Integer>> tagDictionary=new HashMap<Integer, HashSet<Integer>>();
        if(tagDictionaryPath!=null && !tagDictionaryPath.equals("")) {
            BufferedReader tagDictionaryReader = new BufferedReader(new FileReader(tagDictionaryPath));
            while ((line = tagDictionaryReader.readLine()) != null) {
                String[] spl = line.split("\t");
                if (spl.length == 2) {
                    String w = spl[0].toLowerCase();
                    String t = spl[1];

                    if (!stringMap.containsKey(w))
                        stringMap.put(w, index++);
                    if (!stringMap.containsKey(t))
                        stringMap.put(t, index++);

                    int wi = stringMap.get(w);
                    int ti = stringMap.get(t);

                    if (!tagDictionary.containsKey(wi))
                        tagDictionary.put(wi, new HashSet<Integer>());
                    tagDictionary.get(wi).add(ti);
                }
            }
        }
        
        return  new IndexMaps(tagSize,stringMap,reversedMap,clusterNMap,clusterMap,tagDictionary);
    }
}
